### ISSUE TITLE
Adaptive cards transformations for Skype? No activities sent on card action click

### ISSUE THREAD

**Issue by AndreMantas** [Microsoft/BotFramework-Services#70]
I know there are some transformations happening to adaptive cards before being sent to specific channels (like to facebook, where cards appear just as images).
According [to this](https://github.com/Microsoft/botbuilder-dotnet/issues/1202#issuecomment-479574216):
> The BotBuilder SDK does not perform any transformations on Adaptive Cards to accommodate specific channels. Any transformations only occur after the card has reached the connector.

Which makes sense.

So I was thinking if some kind of transformation is being done to adaptive cards before being sent to Skype channel?

Currently I'm trying to receive a simple imBack (since postBack is not supported) when a card button is clicked but on Skype it just doesn't work.
I've set the card action `Data` property to a plain string, which in Webchat makes it send an activity back to the bot with the given text in the `Data` property.

I thought there might be a chance that the connector is forcing the `CardAction.Type` to be always `postBack` instead of checking what is in `Data` to decide if it should be `imBack` (for strings) or `postBack` (for objects) [according to [this](https://stackoverflow.com/a/49638930/2826794)]

Any idea or alternatives to continue using adaptive cards? Our bot must work on multiple channels, Facebook and Webchat are working correctly but not getting imBack nor postBack on Skype is a real limitation.

Related issues:

- https://github.com/Microsoft/AdaptiveCards/issues/2611

- https://github.com/Microsoft/botbuilder-dotnet/issues/1202

**Comment by v-kydela** [comment]
@AndreMantas - Could you provide the Adaptive Card in question? Perhaps there's some way to recreate its functionality in a hero card or carousel of cards.

I will ask around to see if I can find anything out about the Skype connector's treatment of Adaptive Cards.

**Comment by AndreMantas** [comment]
Hi. During conversation, bot presents a carousel of cards (min 1, max 3) to the user so he can choose which one he wants.

Each card looks like this:

![image](https://user-images.githubusercontent.com/7627975/55795285-51415400-5abf-11e9-8fb7-7bdf6c1ec318.png)

Each tittle is different, and my workaround to use imBack instead of postBack would be that on "Choose" button click, client chat would send the title as imBack; the bot would then check to see if the answer is valid.
Could even make this the default behaviour as it works in webchat and facebook channels; just need it to work on Skype

**Comment by v-kydela** [comment]
Okay, my recommendation is to have the bot check the channel and if it's Skype it should do something differently. I have two ideas for that:

1. Look into [receipt cards](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-add-rich-card-attachments?view=azure-bot-service-3.0#add-a-receipt-card-to-a-message), since I think a receipt card does everything you want that card to do. You might even consider switching to receipt cards for your other channels as well if you don't want your bot to do things differently for different channels.
2. Send the Adaptive Card without the buttons, and then send the buttons either as suggested actions in the same message or as a hero card in the next message.

You can try out all kinds of cards using the [CardsBot sample](https://github.com/Microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/06.using-cards). A receipt card looks something like this in the Skype Windows client:

![image](https://user-images.githubusercontent.com/41968495/55831492-0e37ae80-5ac8-11e9-9bf5-b7b6834df8e4.png)

**Comment by AndreMantas** [comment]
Thanks for the reply. I've been thinking about option 2 for a while

This is not a receipt so probably I won't use the receipt card. [Facebook best practices](https://developers.facebook.com/docs/messenger-platform/send-messages/template/receipt/) say to not to use receipt card for anything else.

But then I also face another problem with adaptive cards in Skype, because they seem to be cut after a certain height.

Skype please :(

**Comment by v-kydela** [comment]
You were told in your [other issue](https://github.com/Microsoft/AdaptiveCards/issues/2611) that Adaptive Cards were never officially supported in Skype. I recommend using something other than Adaptive Cards.

1. You can use a receipt card.
2. You can use a hero card and just concatenate the items in the text property.
3. You can send the items as a message outside of a card.

There are many workarounds, but Skype does not support Adaptive Cards so there's not much we can do.

Does this resolve your issue?

**Comment by v-kydela** [comment]
@AndreMantas - Are you still working on this?

**Comment by v-kydela** [comment]
Closing due to inactivity

**Comment by AndreMantas** [comment]
Hi. Sorry for late reply.

What I ended up doing is use HeroCard just for title and price (in subtitle) without any image nor text.

Then I added a new button like "see items" in each card and changed bot code to detect which card items user wants to see and reprompt hero cards carrousel after showing the items list.

Skype only uses 2 lines of Text property if Title and Subtitle are set (4 if one of them is set and 6 if none of them is set). My items could go up to 8 so that was not an option.

I later found another issue with Skype.
In Android, AdaptiveCards just don't render. They render somehow in PC & iOS versions, although they have a very specific height (and are cropped beyond that).

So instead of sending AdaptiveCards in attachments I now convert them to images and send in base64 attachments (works fine for now when it is only one).

Hopefully someday I can revert this all back when Skype fully supports AdaptiveCards in all platforms.

Thanks for your help.
